components/raftstore/src/store/fsm/apply.rs

```diff
@@ -1830,19 +1842,106 @@ where
             {
                 peer.set_id(*peer_id);
             }
-            write_peer_state(kv_wb_mut, &new_region, PeerState::Normal, None)
+            new_regions_map.insert(
+                new_region.get_id(),
+                (
+                    util::find_peer(&new_region, ctx.store_id).unwrap().get_id(),
+                    None,
+                ),
+            );
+            regions.push(new_region);
+        }
+
+        if right_derive {
+            derived.set_start_key(keys.pop_front().unwrap());
+            regions.push(derived.clone());
+        }
+
+        let mut meta = ctx.store_meta.lock().unwrap();
+        for (region_id, val) in new_regions_map.iter_mut() {
+            if val.1.is_some() {
+                continue;
+            }
+            let peer_id = val.0;
+            if let Some(r) = meta.regions.get(region_id) {
+                if util::is_region_initialized(r) {
+                    val.1 = Some(format!("region {:?} has already initialized", r));
+                } else {
+                    // If the region in meta.regions is not initialized, it must exist in meta.pending_create_peers.
+                    let status = meta.pending_create_peers.get_mut(region_id).unwrap();
+                    if *status == (peer_id, CreatePeerStatus::Empty) {
+                        *status = (peer_id, CreatePeerStatus::Split);
+                    } else {
+                        val.1 = Some(format!("status {:?} is not expected", status));
+                    }
+                }
+            } else {
+                // If the region not exist in meta.regions, it may exist in meta.pending_create_peers.
+                // See details in `maybe_create_peer`.
+                meta.pending_create_peers
+                    .insert(*region_id, (peer_id, CreatePeerStatus::Split));
+            }
+        }
+        drop(meta);
+
+        let mut already_exist_regions = Vec::new();
+        for (region_id, val) in new_regions_map.iter_mut() {
+            if val.1.is_some() {
+                continue;
+            }
+            let peer_id = val.0;
+            let region_state_key = keys::region_state_key(*region_id);
+            match ctx
+                .engine
+                .get_msg_cf::<RegionLocalState>(CF_RAFT, &region_state_key)
+            {
+                Ok(None) => (),
+                Ok(Some(state)) => {
+                    already_exist_regions.push((*region_id, peer_id));
+                    val.1 = Some(format!("state {:?} exist in kv engine", state));
+                }
```

Comment:
Could we check `RegionLocalState`  before checking `meta.regions` and `meta.pending_create_peers`? Then we don't need to drop and lock `meta` again and make the logic cleaner.

Reply:
After checking `RegionLocalState`, the other side may create a new peer, apply snapshot then destroy it. That's way adding to `pending_create_peers` before checking `RegionLocalState` both in apply fsm and store fsm.

Reply:
If a region did create and destroy between checking `RegionLocalState` and `pending_create_peers` then split successfully, it can still be destroyed when it tries to connect with other peers. BTW, looks like store fsm adding to `pending_create_peers` before checking `RegionLocalState`.